clang/runtime/dpct-rt/include/dpct/bindless_images.hpp

```diff
@@ -111,12 +126,410 @@ class image_mem_wrapper {
+                               "is not yet supported!");
+    }
+    if (array_index >= _res_img_desc.array_size) {
+      throw std::runtime_error(
+          "Array index requested exceeds total array size in resource!");
+    }
+    if (mip_level >= _res_img_desc.num_levels) {
+      throw std::runtime_error(
+          "Mip level requested exceeds total num of levels in resource!");
+    }
+    if (!_res_img_mem_wrapper_ptr) {
+      throw std::runtime_error(
+          "Resouce is not mapped! "
+          "Resource should be mapped before accessing its memory");
+    }
+
+    return (mip_level == 0)
+               ? _res_img_mem_wrapper_ptr
+               : _res_img_mem_wrapper_ptr->get_mip_level(mip_level);
+  }
+
+  /// Unmap the buf/texture resource memory and get its pointer/mem handle.
+  /// \param [in] q The queue used to unmap the resource with.
+  void unmap_resource(sycl::queue q) {
+    if (_res_is_buffer) {
+      if (!_res_buf_ptr) {
+        throw std::runtime_error("Resource is not mapped! "
+                                 "Resource should be mapped before unmapping.");
+      }
+
+      _res_buf_ptr = nullptr;
+
+      throw std::runtime_error(
+          "Unmapping a buffer resoure is not yet supported!");
+    } else {
+      if (!_res_img_mem_wrapper_ptr) {
+        throw std::runtime_error("Resource is not mapped! "
+                                 "Resource should be mapped before unmapping.");
+      }
+
+      delete _res_img_mem_wrapper_ptr;
+      _res_img_mem_wrapper_ptr = nullptr;
+    }
+  }
+
+protected:
+  /// Init external mem handle for a given resource.
+  external_mem_wrapper_base(
+      sycl::ext::oneapi::experimental::external_mem external_mem,
+      unsigned reg_flags)
+      : _res_external_mem(external_mem), _res_reg_flags(reg_flags) {}
+
+  unsigned _res_reg_flags = 0;
+  bool _res_is_buffer = false;
+  size_t _res_size_bytes = 0;
+
+  sycl::ext::oneapi::experimental::image_descriptor _res_img_desc;
+  sycl::ext::oneapi::experimental::external_mem _res_external_mem;
+
+private:
+  void *_res_buf_ptr = nullptr;
+  image_mem_wrapper *_res_img_mem_wrapper_ptr = nullptr;
+};
+
+#ifdef _WIN32
+class external_mem_wrapper : public external_mem_wrapper_base {
+public:
+  /// Creates external memory wrapper for a given DX11 resource.
+  /// \param [in] d3d11_res Pointer to the ID3D11Resource to be wrapped.
+  /// \param [in] reg_flags Registration flags used for resource registration.
+  external_mem_wrapper(ID3D11Resource *d3d11_res, unsigned reg_flags)
+      : _res_D3D11(d3d11_res),
+        external_mem_wrapper_base(
+            sycl::ext::oneapi::experimental::external_mem(), reg_flags) {
+    // Check if the resource is already registered to prevent double
+    // registration.
+    if (_res_D3D11_set.find(_res_D3D11) != _res_D3D11_set.end()) {
+      throw std::runtime_error("This D3D11 resource is already registered!");
```

Comment:
can we return the registered one directly without runtime_error?

Reply:
This logic is added for emulating the CUDA behavior here ([ref](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__D3D11.html#group__CUDART__D3D11_1g85d07753780643584b8febab0370623b:~:text=If%20pD3DResource%20is%20of%20incorrect%20type%20or%20is%20already%20registered%2C%20then%20cudaErrorInvalidResourceHandle%20is%20returned))